pennylane/fourier/advanced_spectrum.py

```diff
@@ -0,0 +1,365 @@
+
+    >>> print(res)
+    {"x": {(0,): [-0.4, 0.0, 0.4], (1,): [-3.14159, 0.0, 3.14159]}}
+
+    """
+    validation_kwargs = validation_kwargs or {}
+    encoding_args, argnum = _process_ids(encoding_args, argnum, qnode)
+    atol = 10 ** (-decimals) if decimals is not None else 1e-10
+    # A map between Jacobians (contiguous) and arg names (may be discontiguous)
+    arg_name_map = dict(enumerate(encoding_args))
+
+    @wraps(qnode)
+    def wrapper(*args, **kwargs):
+        # Compute classical Jacobian and assert preprocessing is linear
+        jac_fn = qml.transforms.classical_jacobian(qnode, argnum=argnum, expand_fn=expand_multi_par)
+        if not qml.math.is_independent(jac_fn, qnode.interface, args, kwargs, **validation_kwargs):
+            raise ValueError(
+                "The Jacobian of the classical preprocessing in the provided QNode "
+                "is not constant; only linear classical preprocessing is supported."
+            )
+        class_jacs = jac_fn(*args, **kwargs)
+
+        spectra = {}
+        tape = expand_multi_par(qnode.qtape)
+        par_info = tape._par_info  # pylint: disable=protected-access
+
+        for jac_idx, class_jac in enumerate(class_jacs):
+            arg_name = arg_name_map[jac_idx]
+            if encoding_args[arg_name] is Ellipsis:
+                requested_par_ids = set(product(*(range(sh) for sh in class_jac.shape[1:])))
+            else:
+                requested_par_ids = set(encoding_args[arg_name])
+            _spectra = {par_idx: {0} for par_idx in requested_par_ids}
+
+            for op_idx, jac_of_op in enumerate(np.round(class_jac, decimals=decimals)):
+                op = par_info[op_idx]["op"]
+                # Find parameters that where requested and feed into the operation
+                if len(class_jac.shape) == 1:
+                    # Scalar argument, only axis of Jacobian is for gates
+                    if np.isclose(jac_of_op, 0.0, atol=atol, rtol=0):
+                        continue
+                    jac_of_op = {(): jac_of_op}
+                    par_ids = {()}
+                else:
+                    par_ids = zip(*[map(int, _ids) for _ids in np.where(jac_of_op)])
+                    par_ids = set(par_ids).intersection(requested_par_ids)
+                    if len(par_ids) == 0:
+                        continue
+                # Multi-parameter gates are not supported
+                if len(op.parameters) != 1:
+                    raise ValueError(
+                        "Can only consider one-parameter gates as data-encoding gates; "
+                        f"got {op.name}."
+                    )
+                spec = _get_spectrum(op, decimals=decimals)
+
+                for par_idx in par_ids:
+                    scale = float(jac_of_op[par_idx])
+                    scaled_spec = [scale * f for f in spec]
+                    _spectra[par_idx] = _join_spectra(_spectra[par_idx], scaled_spec)
```

Comment:
@dwierichs I had a bit of trouble following this function, it is quite dense! Would be nice to add more detailed commenting to it, explaining each branch 🙂

Reply:
I hope it's better now :)